File: app/src/main/java/com/argos/android/opencv/Adapter/FeatureListAdapter.java
package com.argos.android.opencv.Adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import com.argos.android.opencv.Interface.MainActivityCallback;
import com.argos.android.opencv.Model.Feature;
import com.argos.android.opencv.R;
import java.util.ArrayList;

/**
 * Adapter for the cards on the first screen
 */

public class FeatureListAdapter extends RecyclerView.Adapter<FeatureListAdapter.ViewHolder>
{
    private ArrayList<Feature> features;
    private LayoutInflater inflater;
    private MainActivityCallback mainActivityCallback;

    public FeatureListAdapter(ArrayList<Feature> features, Context context)
    {
        this.features = features;
        inflater = LayoutInflater.from(context);
        mainActivityCallback = (MainActivityCallback) context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
    {
        return new ViewHolder(inflater.inflate(R.layout.feature_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, final int position)
    {
        holder.featureName.setText(features.get(position).getFeatureName());
        holder.featureThumbnail.setImageResource(features.get(position).getFeatureThumbnail());

        holder.pictureButton.setOnClickListener(
                new View.OnClickListener()
                {
                    @Override
                    public void onClick(View v)
                    {
                        mainActivityCallback.chooseImage(features.get(position));
                    }
                }
        );

        holder.cameraButton.setOnClickListener(
                new View.OnClickListener()
                {
                    @Override
                    public void onClick(View v)
                    {
                        mainActivityCallback.launchCameraDetection(features.get(position));
                    }
                }
        );
    }

    @Override
    public int getItemCount()
    {
        return features.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder
    {
        TextView featureName;
        ImageView featureThumbnail;
        ImageButton pictureButton;
        ImageButton cameraButton;

        public ViewHolder(View itemView)
        {
            super(itemView);
            featureName = (TextView) itemView.findViewById(R.id.feature_name);
            featureThumbnail = (ImageView) itemView.findViewById(R.id.feature_thumbnail);
            pictureButton = (ImageButton) itemView.findViewById(R.id.button_picture);
            cameraButton = (ImageButton) itemView.findViewById(R.id.button_camera);
        }
    }
}

File: app/src/main/java/com/argos/android/opencv/adapter/FeatureListAdapter.kt
package com.argos.android.opencv.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import com.argos.android.opencv.interfaces.MainActivityCallback
import com.argos.android.opencv.model.Feature
import com.argos.android.opencv.R
import java.util.*

/**
 * Adapter for the cards on the first screen
 */

class FeatureListAdapter(private val features: ArrayList<Feature>, context: Context) : RecyclerView.Adapter<FeatureListAdapter.ViewHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val mainActivityCallback: MainActivityCallback = context as MainActivityCallback

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(inflater.inflate(R.layout.feature_list_item, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.featureName.text = features[position].featureName
        holder.featureThumbnail.setImageResource(features[position].featureThumbnail)

        holder.pictureButton.setOnClickListener { mainActivityCallback.chooseImage(features[position]) }

        holder.cameraButton.setOnClickListener { mainActivityCallback.launchCameraDetection(features[position]) }
    }

    override fun getItemCount(): Int {
        return features.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var featureName: TextView = itemView.findViewById<View>(R.id.feature_name) as TextView
        var featureThumbnail: ImageView = itemView.findViewById<View>(R.id.feature_thumbnail) as ImageView
        var pictureButton: ImageButton = itemView.findViewById<View>(R.id.button_picture) as ImageButton
        var cameraButton: ImageButton = itemView.findViewById<View>(R.id.button_camera) as ImageButton
    }
}
